test/jasmine/tests/mock_test.js

```diff
@@ -0,0 +1,2081 @@
+figs['violin_only_zeroes'] = require('@mocks/violin_only_zeroes');
+figs['violin_positive_and_negative'] = require('@mocks/violin_positive_and_negative');
+figs['violin_positive_sides_w_points'] = require('@mocks/violin_positive_sides_w_points');
+figs['violin_ridgeplot'] = require('@mocks/violin_ridgeplot');
+figs['violin_side-by-side'] = require('@mocks/violin_side-by-side');
+figs['violin_style'] = require('@mocks/violin_style');
+figs['violin_zoomed-in'] = require('@mocks/violin_zoomed-in');
+figs['violin-offsetgroups'] = require('@mocks/violin-offsetgroups');
+figs['viridis_heatmap'] = require('@mocks/viridis_heatmap');
+figs['waterfall_11'] = require('@mocks/waterfall_11');
+figs['waterfall_and_bar'] = require('@mocks/waterfall_and_bar');
+figs['waterfall_and_histogram'] = require('@mocks/waterfall_and_histogram');
+figs['waterfall_attrs'] = require('@mocks/waterfall_attrs');
+figs['waterfall_axis'] = require('@mocks/waterfall_axis');
+figs['waterfall_cliponaxis-false'] = require('@mocks/waterfall_cliponaxis-false');
+figs['waterfall_custom'] = require('@mocks/waterfall_custom');
+figs['waterfall_date-axes'] = require('@mocks/waterfall_date-axes');
+figs['waterfall_funnel_texttemplate_date'] = require('@mocks/waterfall_funnel_texttemplate_date');
+figs['waterfall_gap0'] = require('@mocks/waterfall_gap0');
+figs['waterfall_line'] = require('@mocks/waterfall_line');
+figs['waterfall_months'] = require('@mocks/waterfall_months');
+figs['waterfall_multicategory'] = require('@mocks/waterfall_multicategory');
+figs['waterfall_nonnumeric_sizes'] = require('@mocks/waterfall_nonnumeric_sizes');
+figs['waterfall_profit-loss_2018_positive-negative'] = require('@mocks/waterfall_profit-loss_2018_positive-negative');
+figs['waterfall_profit-loss_2018vs2019_overlay'] = require('@mocks/waterfall_profit-loss_2018vs2019_overlay');
+figs['waterfall_profit-loss_2018vs2019_rectangle'] = require('@mocks/waterfall_profit-loss_2018vs2019_rectangle');
+figs['waterfall_profit-loss_2018vs2019_textinfo_base'] = require('@mocks/waterfall_profit-loss_2018vs2019_textinfo_base');
+figs['waterfall-grouping-vs-defaults'] = require('@mocks/waterfall-grouping-vs-defaults');
+figs['waterfall-offsetgroups'] = require('@mocks/waterfall-offsetgroups');
+figs['world-cals'] = require('@mocks/world-cals');
+figs['yaxis-over-yaxis2'] = require('@mocks/yaxis-over-yaxis2');
+figs['yignbu_heatmap'] = require('@mocks/yignbu_heatmap');
+figs['yiorrd_heatmap'] = require('@mocks/yiorrd_heatmap');
+figs['zsmooth_methods'] = require('@mocks/zsmooth_methods');
+*/
+
+describe('@noCI mock validation', function() {
+    list.forEach(function(name) {
+        var figure = figs[name];
+        if(figure) {
+            it('validating mock: "' + name + '"', function() {
+                var out = Plotly.validate(
+                    figure.data,
+                    figure.layout || {} // shouldn't Plotly.validate handle undefined layout?
```

Comment:
Yes, I think you're right, it should. Or even undefined data, like if all you want are shapes and annotations. Feel free to make a new issue for this, or just fix it in this PR if you like.

Reply:
Good point.
New issue: https://github.com/plotly/plotly.js/issues/4797.
Resolving...